README
dompdf - PHP5 HTML to PDF converter
===================================

http://www.digitaljunkies.ca/dompdf
Copyright (c) 2004-2005 Benj Carson
R&OS PDF class (class.pdf.php) Copyright (c) 2001-04 Wayne Munro

Send bug reports, patches, feature requests, complaints & hate mail (no spam
thanks) to <benjcarson@digitaljunkies.ca>

##### See INSTALL for installation instructions. #####


Table of Contents:

1. Overview
2. Features
3. Requirements
4. Limitations (Known Issues)
5. Usage
6. Inline PHP Support

Overview
--------

dompdf is an HTML to PDF converter.  At its heart, dompdf is (mostly)
CSS2.1 compliant HTML layout and rendering engine written in PHP.  It is
a style-driven renderer: it will download and read external stylesheets,
inline style tags, and the style attributes of individual HTML elements.  It
also supports most presentational HTML attributes.

PDF rendering is currently provided either by PDFLib (www.pdflib.com)
or by a bundled version the R&amp;OS CPDF class written by Wayne Munro
(www.ros.co.nz/pdf).  (Some performance related changes have been made
to the R&amp;OS class, however).  In order to use PDFLib with dompdf,
the PDFLib PECL extension is required.  Using PDFLib improves
performance and reduces the memory requirements of dompdf somewhat,
while the R&amp;OS CPDF class, though slightly slower, eliminates any
dependencies on external PDF libraries.

dompdf was entered in the Zend PHP 5 Contest and placed 20th overall.

Please note that dompdf works only with PHP 5.  There are no plans for
a PHP 4 port.  If your web host does not offer PHP 4, I suggest either pestering
them, or setting up your own PHP 5 box and using it to run dompdf.  Your scripts
on your web host can redirect PDF requests to your PHP 5 box.

This package should contain:

dompdf.php                 PDF Generating script
dompdf_config.inc.php      Main configuration file
load_font.php              Font loading utility script
HACKING                    Notes on messing with the code
INSTALL                    Installation instructions
LICENSE.LGPL               GNU Lesser General Public License
NEWS                       Release news
README                     This file
TODO                       Things I'm working on
include/                   PHP class & include files
lib/                       R&OS PDF class, fonts, default CSS file
www/                       Demonstration webpage
www/test/                  Some test HTML pages


For the impatient:

Once you have installed dompdf, point your browser at the www/ directory
for HTML documentation and a quick demonstration.


Features
--------

* handles most CSS2.1 properties, including @import, @media & @page rules

* supports most presentational HTML 4.0 attributes

* supports external stylesheets, either local or through http/ftp (via
  fopen-wrappers)

* supports complex tables, including row & column spans, separate &
  collapsed border models, individual cell styling, (no nested tables yet
  however)

* image support (gif, png & jpeg)

* no dependencies on external PDF libraries, thanks to the R&OS PDF class

* inline PHP support.  See below for details.


Requirements
------------

* PHP 5.0.0+

* Some fonts.  PDFs internally support Helvetica, Times-Roman, Courier &
  Zapf-Dingbats, but if you wish to use other fonts you will need to install
  some fonts.  dompdf supports the same fonts as the underlying R&OS PDF
  class: Type 1 (.pfb with the corresponding .afm) and TrueType (.ttf).  At
  the minimum, you should probably have the Microsoft core fonts (now
  available at: http://corefonts.sourceforge.net/).  See the INSTALL file
  for font installation instructions.


Limitations (Known Issues)
--------------------------

* tables can not be nested

* not particularly tolerant to poorly-formed HTML input (using Tidy first
  may help).

* large files can take a while to render

* ordered lists are currently not supported

Usage
-----

The included dompdf.php script can be used both from the command line or via
a web browser.  Alternatively, the dompdf class can be used directly.

Invoking dompdf via the web:

The dompdf.php script is not intended to be an interactive page.  It
receives input parameters via $_GET and can stream a PDF directly to the
browser.  This makes it possible to embed links to the script in a page that
look like static PDF links, but are actually dynamically generated.  This
method is also useful as a redirection target.

dompdf.php accepts the following $_GET variables:

input_file   required    a rawurlencoded() path to the HTML file to
                         process.  Remote files (http/ftp) are supported if
                         fopen wrappers are enabled.

paper        optional    the paper size.  Defaults to 'letter' (unless the
                         default has been changed in dompdf_config.inc.php).
                         See include/cpdf_adapter.cls.php, or invoke
                         dompdf.php on the command line with the -l switch
                         for accepted paper sizes.

orientation  optional    'portrait' or 'landscape'.  Defaults to 'portrait'.

base_path    optional    the base path to use when resolving relative links 
                         (images or CSS files).  Defaults to the directory
                         containing the file being accessed.  (This option is
                         useful for pointing dompdf at your CSS files even
                         though the HTML file may be elsewhere.)

output_file  optional    the rawurlencoded() name of the output file.
                         Defaults to 'dompdf_out.pdf'.

save_file    optional    If present (i.e. isset($_GET["save_file"]) ==
                         true), output_file is saved locally,  Otherwise
                         the file is streamed directly to the client.


One technique for generating dynamic PDFs is to generate dynamic HTML as you
normally would, except instead of displaying the output to the browser, you
use output buffering and write the output to a temporary file.  Once this
file is saved, you redirect to the dompdf.php script.  If you use a
templating engine like Smarty, you can simply do:

<?php
$tmpfile = tempnam("/tmp", "dompdf_");
file_put_contents($tmp_file, $smarty->fetch());

$url = "dompdf.php?input_file=" . rawurlencode($tmpfile) . 
       "&paper=letter&output_file=" . rawurlencode("My Fancy PDF.pdf");

header("Location: http://" . $_SERVER["HTTP_HOST"] . "/$url");
?>

If you use any stylesheets, you may need to provide the base_path option to
tell dompdf where to look for them, as they are not likely relative to 
/tmp ;).


Invoking dompdf via the command line:

You can execute dompdf.php using the following command:

$ php -f dompdf.php -- [options]

(If you find yourself using only the cli interface, you can add
#!/usr/bin/php as the first line of dompdf.php to invoke dompdf.php
directly.)

dompdf.php is invoked as follows:

$ ./dompdf.php [options] html_file
  
  html_file can be a filename, a url if fopen_wrappers are enabled, or the
  '-' character to read from standard input.

  -h             Show a brief help message

  -l             list available paper sizes

  -p size        paper size; something like 'letter', 'A4', 'legal', etc.  
                 Thee default is 'letter'

  -o orientation either 'portrait' or 'landscape'.  Default is 'portrait'.

  -b path        the base path to use when resolving relative links 
                 (images or CSS files). Default is the directory of
                 html_file.

  -f file        the output filename.  Default is the input [html_file].pdf.

  -v             verbose: display html parsing warnings and file not found 
                 errors.

  -d             very verbose: display oodles of debugging output; every 
                 frame in the tree is printed to stdout.

Examples:

$ php -f dompdf.php -- my_resume.html
$ php -f dompdf.php -- -b /var/www/ ./web_stuff/index.html
$ echo '<html><body>Hello world!</body>' | php -f dompdf.php -- -


Using the dompdf class directly:

See the API documentation for the interface definition.  The API
documentation is available at http://www.digitaljunkies.ca/dompdf/.  



Inline PHP Support
------------------

dompdf supports two varieties of inline PHP code.  All PHP evaluation is
controlled by the DOMPDF_ENABLE_PHP configuration option.  If it is set to
false, then no PHP code is executed.  Otherwise, PHP is evaluated in two
passes:

The first pass is useful for inserting dynamic data into your PDF.  You can
do this by embedding <?php ?> tags in your HTML file, as you would in a
normal .php file.  This code is evaluated prior to parsing the HTML, so you
can echo any text or markup and it will appear in the rendered PDF.

The second pass is useful for performing drawing operations on the
underlying PDF class directly.  You can do this by embedding PHP code within
<script type="text/php"> </script> tags.  This code is evaluated during the
rendering phase and you have access to a few internal objects and
operations.  In particular, the $pdf variable is the current instance of
CPDF_Adapter.  Using this object, you can write and draw directly on the
current page.  Using the CPDF_Adapter::open_object(),
CPDF_Adapter::close_object() and CPDF_Adapter::add_object() methods, you can
create text and drawing objects that appear on every page of your PDF
(useful for headers & footers).

The following variables are defined for you during the second pass of PHP
execution:

  $pdf         the current instance of CPDF_Adapter
  $PAGE_NUM    the current page number
  $PAGE_COUNT  the total number of pages in the document

For more complete documentation of the CPDF_Adapter API, see either
include/cpdf_adapter.cls.php and include/canvas.cls.php directly, or check
out the online documentation at http://www.digitaljunkies.ca/dompdf/doc

That's it!  Have fun!


Send questions, problems, bug reports, etc to:

Benj Carson <benjcarson@digitaljunkies.ca>


ChangeLog
2007-08-22 16:02  benjcarson

* include/block_frame_decorator.cls.php,
  include/block_frame_reflower.cls.php, include/frame.cls.php,
  include/functions.inc.php, include/inline_frame_decorator.cls.php,
  include/inline_renderer.cls.php,
  include/table_cell_frame_reflower.cls.php,
  include/text_frame_decorator.cls.php, lib/class.pdf.php:

  - Absolute positioning fixes from Greg Nakatsui:
   - Improve font height calculation in cpdf class
   - Adjust vertical spacing in tds
   - Fix border rendering on inline elements

2007-06-24 19:45  benjcarson

* dompdf.php, dompdf_config.inc.php, include/absolute_positioner.cls.php,
  include/block_frame_reflower.cls.php, include/cellmap.cls.php,
  include/frame.cls.php, include/frame_factory.cls.php,
  include/frame_tree.cls.php, include/functions.inc.php,
  include/table_cell_frame_reflower.cls.php,
  include/table_frame_reflower.cls.php,
  include/table_row_frame_reflower.cls.php,
  include/table_row_group_frame_reflower.cls.php:

  - Additional fixes for "Frame not found in cellmap" errors by Greg Nakatsui
  - inital absolute positioning implementation

2006-12-22 10:37  benjcarson

* dompdf_config.inc.php, include/page_frame_decorator.cls.php,
  include/renderer.cls.php, include/table_frame_decorator.cls.php:

  - render list item backgrounds and borders
  - fix "nesting level too deep" errors on PHP 5.2

2006-12-14 06:36  benjcarson

* dompdf_config.inc.php:

  - added check to  __autoload to improve compatibility with other apps

2006-10-26 10:07  benjcarson

* ChangeLog, include/attribute_translator.cls.php,
  include/cpdf_adapter.cls.php,
  include/list_bullet_frame_decorator.cls.php,
  include/list_bullet_renderer.cls.php, www/faq.php, www/usage.php:

  - keep attribute translator from changing borders on TDs if they have already been specified
  - ul adjustments

2006-10-18 14:49  benjcarson

* include/stylesheet.cls.php:

  - added support for :first-child pseudo-selector

2006-10-13 16:14  benjcarson

* include/: block_frame_decorator.cls.php, text_frame_decorator.cls.php,
  text_frame_reflower.cls.php:

  - improved fix for collapsing white space bug

2006-10-12 15:02  benjcarson

* dompdf.php, include/block_frame_decorator.cls.php,
  include/dompdf.cls.php, include/frame.cls.php, include/functions.inc.php,
  include/generated_frame_reflower.cls.php,
  include/inline_frame_reflower.cls.php, include/renderer.cls.php,
  include/style.cls.php, include/stylesheet.cls.php,
  include/text_frame_reflower.cls.php, include/text_renderer.cls.php,
  www/faq.php:

  - fixed problem with missing whitespace in inline elements

2006-10-11 12:52  benjcarson

* dompdf_config.inc.php, include/cpdf_adapter.cls.php,
  include/list_bullet_frame_decorator.cls.php,
  include/list_bullet_frame_reflower.cls.php,
  include/list_bullet_renderer.cls.php, test/ul.html:

  - fixes and adjustments to ul handling

2006-10-03 15:04  benjcarson

* ChangeLog, Makefile, dompdf.php, dompdf_config.inc.php,
  include/attribute_translator.cls.php, www/faq.php:

  - fixed issue with vspace and hspace img attributes

2006-09-29 10:37  benjcarson

* include/attribute_translator.cls.php:

  - fix for bug #1534006: Borders set via attributes not applied to TH elements

2006-09-19 11:24  benjcarson

* include/dompdf.cls.php:

  - fixed security issue identified by Alex Günsche:
    - prevent opening of dot-files

2006-09-19 11:20  benjcarson

* dompdf.php, dompdf_config.inc.php, include/dompdf.cls.php,
  include/style.cls.php:

  - fixed security issue identified by Alex Günsche:
    - added DOMPDF_CHROOT configuration variable to restrict input file location

2006-08-11 11:04  benjcarson

* dompdf_config.inc.php, load_font.php, include/functions.inc.php,
  include/gd_adapter.cls.php, include/page_frame_decorator.cls.php,
  include/pdflib_adapter.cls.php, include/style.cls.php,
  include/table_frame_reflower.cls.php,
  include/text_frame_reflower.cls.php, lib/class.pdf.php:

  - fix for local image paths (command-line only?)
  - support for text-transform

2006-08-02 11:44  benjcarson

* dompdf.php, dompdf_config.inc.php, include/cellmap.cls.php,
  include/frame_tree.cls.php, include/functions.inc.php,
  include/image_cache.cls.php, include/image_frame_decorator.cls.php,
  include/page_frame_decorator.cls.php,
  include/table_cell_frame_reflower.cls.php,
  include/table_frame_reflower.cls.php,
  include/table_row_frame_reflower.cls.php:

  - fixed various frame not found in cellmap errors
  - improved support for nested tables breaking across pages

2006-07-21 14:23  benjcarson

* ChangeLog, include/cellmap.cls.php, include/cpdf_adapter.cls.php,
  include/frame_reflower.cls.php, include/functions.inc.php,
  include/table_frame_decorator.cls.php,
  include/table_frame_reflower.cls.php,
  include/text_frame_decorator.cls.php, test/nested_table.html:

  - initial official support for nested tables.
  - fixed table height when last row is moved to new page, fixes table background & border being offset from table

2006-07-07 14:33  benjcarson

* test/: anchor_link.html, background_image.html, baseline.html,
  block_height.html, border_css_values.html, br.html, center_table.html,
  gif_test.html, multiple_class.html, nbsp.html, table_height.html,
  table_image.html:

  - added missing tests to cvs

2006-07-07 14:31  tag dompdf-0-5-1

2006-07-07 14:31  benjcarson

* ChangeLog, Makefile, dompdf.php, dompdf_config.inc.php, load_font.php,
  include/abstract_renderer.cls.php, include/attribute_translator.cls.php,
  include/block_frame_decorator.cls.php,
  include/block_frame_reflower.cls.php, include/block_positioner.cls.php,
  include/block_renderer.cls.php, include/cached_pdf_decorator.cls.php,
  include/canvas.cls.php, include/canvas_factory.cls.php,
  include/cellmap.cls.php, include/cpdf_adapter.cls.php,
  include/dompdf.cls.php, include/dompdf_exception.cls.php,
  include/dompdf_internal_exception.cls.php, include/font_metrics.cls.php,

dompdf.php
<?php
/**
 * DOMPDF - PHP5 HTML to PDF renderer
 *
 * File: $RCSfile: dompdf.php,v $
 * Created on: 2004-06-22
 *
 * Copyright (c) 2004 - Benj Carson <benjcarson@digitaljunkies.ca>
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this library in the file LICENSE.LGPL; if not, write to the
 * Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA
 * 02111-1307 USA
 *
 * Alternatively, you may distribute this software under the terms of the
 * PHP License, version 3.0 or later.  A copy of this license should have
 * been distributed with this file in the file LICENSE.PHP .  If this is not
 * the case, you can obtain a copy at http://www.php.net/license/3_0.txt.
 *
 * The latest version of DOMPDF might be available at:
 * http://www.digitaljunkies.ca/dompdf
 *
 * dompdf.php is a simple script to drive DOMPDF.  It can be executed from
 * a browser or from the command line.
 *
 * @link http://www.digitaljunkies.ca/dompdf
 * @copyright 2004 Benj Carson
 * @author Benj Carson <benjcarson@digitaljunkies.ca>
 * @package dompdf
 * @version 0.5.1
 */

/* $Id: dompdf.php,v 1.22 2007-06-25 02:45:11 benjcarson Exp $ */

/**
 * Display command line usage:
 *
 * Usage: ./dompdf.php [options] html_file
 *
 * html_file can be a filename, a url if fopen_wrappers are enabled, or the '-'
 * character to read from standard input.
 *
 * Options:
 *  -h             Show this message
 *  -l             list available paper sizes
 *  -p size        paper size; something like 'letter', 'A4', 'legal', etc.  The default is
 *                 'letter'
 *  -o orientation either 'portrait' or 'landscape'.  Default is 'portrait'.
 *  -b path        set the 'document root' of the html_file.  Relative urls (for
 *                 stylesheets) are resolved using this directory.  Default is the
 *                 directory of html_file.
 *  -f file        the output filename.  Default is the input [html_file].pdf.
 *  -v             verbose: display html parsing warnings and file not found errors.
 *  -d             very verbose: display oodles of debugging output: every frame in the
 *                 tree is printed to stdout.
 *  -t             comma separated list of debugging types (page-break,reflow,split)
 *  -r             write the render time to the log file
 *
 *
 */

function dompdf_usage() {
  echo
    "\nUsage: {$_SERVER["argv"][0]} [options] html_file\n\n".
    "html_file can be a filename, a url if fopen_wrappers are enabled, or the '-' \n".
    "character to read from standard input.\n\n".
    "Options:\n".
    " -h\t\tShow this message\n".
    " -l\t\tlist available paper sizes\n".
    " -p size\tpaper size; something like 'letter', 'A4', 'legal', etc.  The default is\n".
    "   \t\t'" . DOMPDF_DEFAULT_PAPER_SIZE . "'\n".
    " -o orientation\teither 'portrait' or 'landscape'.  Default is 'portrait'.\n".
    " -b path\tset the 'document root' of the html_file.  Relative urls (for \n".
    "        \tstylesheets) are resolved using this directory.  Default is the \n".
    "        \tdirectory of html_file.\n".
    " -f file\tthe output filename.  Default is the input [html_file].pdf.\n".
    " -v     \tverbose: display html parsing warnings and file not found errors.\n".
    " -d     \tvery verbose:  display oodles of debugging output: every frame\n".
    "        \tin the tree printed to stdout.\n".
    " -t             comma separated list of debugging types (page-break,reflow,split)\n\n";
    

}

function getoptions() {

  $opts = array();

  if ( $_SERVER["argc"] == 1 )
    return $opts;

  $i = 1;
  while ($i < $_SERVER["argc"]) {

    switch ($_SERVER["argv"][$i]) {

    case "--help":
    case "-h":
      $opts["h"] = true;
      $i++;
      break;

    case "-l":
      $opts["l"] = true;
      $i++;
      break;

    case "-p":
      if ( !isset($_SERVER["argv"][$i+1]) )
        die("-p switch requires a size parameter\n");
      $opts["p"] = $_SERVER["argv"][$i+1];
      $i += 2;
      break;

    case "-o":
      if ( !isset($_SERVER["argv"][$i+1]) )
        die("-o switch requires an orientation parameter\n");
      $opts["o"] = $_SERVER["argv"][$i+1];
      $i += 2;
      break;

    case "-b":
      if ( !isset($_SERVER["argv"][$i+1]) )
        die("-b switch requires a path parameter\n");
      $opts["b"] = $_SERVER["argv"][$i+1];
      $i += 2;
      break;

    case "-f":
      if ( !isset($_SERVER["argv"][$i+1]) )
        die("-f switch requires a filename parameter\n");
      $opts["f"] = $_SERVER["argv"][$i+1];
      $i += 2;
      break;

    case "-v":
      $opts["v"] = true;
      $i++;
      break;

    case "-d":
      $opts["d"] = true;
      $i++;
      break;

    case "-t":
      if ( !isset($_SERVER['argv'][$i + 1]) )
        die("-t switch requires a comma separated list of types\n");
      $opts["t"] = $_SERVER['argv'][$i+1];
      $i += 2;
      break;

   default:
      $opts["filename"] = $_SERVER["argv"][$i];
      $i++;
      break;
    }

  }
  return $opts;
}

require_once("dompdf_config.inc.php");
global $_dompdf_show_warnings;
global $_dompdf_debug;
global $_DOMPDF_DEBUG_TYPES;

$old_limit = ini_set("memory_limit", "80M");

$sapi = php_sapi_name();

switch ( $sapi ) {

 case "cli":

  $opts = getoptions();

  if ( isset($opts["h"]) || (!isset($opts["filename"]) && !isset($opts["l"])) ) {
    dompdf_usage();
    exit;
  }

  if ( isset($opts["l"]) ) {
    echo "\nUnderstood paper sizes:\n";

    foreach (array_keys(CPDF_Adapter::$PAPER_SIZES) as $size)
      echo "  " . mb_strtoupper($size) . "\n";
    exit;
  }
  $file = $opts["filename"];

  if ( isset($opts["p"]) )
    $paper = $opts["p"];
  else
    $paper = DOMPDF_DEFAULT_PAPER_SIZE;

  if ( isset($opts["o"]) )
    $orientation = $opts["o"];
  else
    $orientation = "portrait";

  if ( isset($opts["b"]) )
    $base_path = $opts["b"];

  if ( isset($opts["f"]) )
    $outfile = $opts["f"];
  else {
    if ( $file == "-" )
      $outfile = "dompdf_out.pdf";
    else
      $outfile = str_ireplace(array(".html", ".htm", ".php"), "", $file) . ".pdf";
  }

  if ( isset($opts["v"]) )
    $_dompdf_show_warnings = true;

  if ( isset($opts["d"]) ) {
    $_dompdf_show_warnings = true;
    $_dompdf_debug = true;
  }

  if ( isset($opts['t']) ) {
    $arr = split(',',$opts['t']);
    $types = array();
    foreach ($arr as $type)
      $types[ trim($type) ] = 1;
    $_DOMPDF_DEBUG_TYPES = $types;
  }
  
  $save_file = true;

  break;

 default:

   if ( isset($_GET["input_file"]) )
     $file = rawurldecode($_GET["input_file"]);
   else
     throw new DOMPDF_Exception("An input file is required (i.e. input_file _GET variable).");

   if ( isset($_GET["paper"]) )
     $paper = rawurldecode($_GET["paper"]);
   else
     $paper = DOMPDF_DEFAULT_PAPER_SIZE;

   if ( isset($_GET["orientation"]) )
     $orientation = rawurldecode($_GET["orientation"]);
   else
     $orientation = "portrait";

   if ( isset($_GET["base_path"]) )
     $base_path = rawurldecode($_GET["base_path"]);

   if ( isset($_GET["output_file"]) )
     $outfile = rawurldecode($_GET["output_file"]);
   else
     $outfile = "dompdf_out.pdf";

   if ( isset($_GET["save_file"]) )
     $save_file = true;
   else
     $save_file = false;

   break;
}

$dompdf = new DOMPDF();

if ( $file == "-" ) {
  $str = "";
  while ( !feof(STDIN) )
    $str .= fread(STDIN, 4096);

  $dompdf->load_html($str);

} else
  $dompdf->load_html_file($file);

if ( isset($base_path) ) {
  $dompdf->set_base_path($base_path);
}

$dompdf->set_paper($paper, $orientation);

$dompdf->render();

if ( $_dompdf_show_warnings ) {
  foreach ($_dompdf_warnings as $msg)
    echo $msg . "\n";
  echo $dompdf->get_canvas()->messages;
  flush();
}

if ( $save_file ) {
//   if ( !is_writable($outfile) )
//     throw new DOMPDF_Exception("'$outfile' is not writable.");
  if ( strtolower(DOMPDF_PDF_BACKEND) == "gd" )
    $outfile = str_replace(".pdf", ".png", $outfile);

  list($proto, $host, $path, $file) = explode_url($outfile);
  if ( $proto != "" ) // i.e. not file://
    $outfile = $file; // just save it locally, FIXME? could save it like wget: ./host/basepath/file

  $outfile = dompdf_realpath($outfile);

  if ( strpos($outfile, DOMPDF_CHROOT) !== 0 )
    throw new DOMPDF_Exception("Permission denied.");

  file_put_contents($outfile, $dompdf->output());
  exit(0);
}

if ( !headers_sent() ) {
  $dompdf->stream($outfile);
}
?>